diff/reader_util.go

```diff
@@ -2,9 +2,77 @@ package diff
 
 import (
 	"bufio"
+	"bytes"
 	"io"
 )
 
+func newLineReader(r io.Reader) *lineReader {
+	return &lineReader{reader: bufio.NewReader(r)}
+}
+
+// lineReader is a wrapper around a bufio.Reader that caches the next line to
+// provide lookahead functionality for the next two lines.
+type lineReader struct {
+	reader *bufio.Reader
+
+	cachedNextLine    []byte
+	cachedNextLineErr error
+}
+
+// readLine returns the next unconsumed line and advances the internal cache of
+// the lineReader.
+func (l *lineReader) readLine() ([]byte, error) {
+	if l.cachedNextLine == nil && l.cachedNextLineErr == nil {
+		l.cachedNextLine, l.cachedNextLineErr = readLine(l.reader)
+	}
+
+	if l.cachedNextLineErr != nil {
+		return nil, l.cachedNextLineErr
+	}
+
+	next := l.cachedNextLine
+
+	l.cachedNextLine, l.cachedNextLineErr = readLine(l.reader)
+
+	return next, nil
+}
+
+// nextLineStartsWith looks at the line that would be returned by the next call
+// to readLine to check whether it has the given prefix.
+//
+// io.EOF and bufio.ErrBufferFull errors are ignored so that the function can
+// be used when at the end of the file.
+func (l *lineReader) nextLineStartsWith(prefix string) (bool, error) {
+	return l.lineHasPrefix(l.cachedNextLine, prefix, l.cachedNextLineErr)
```

Comment:
if someone was to use it before using readLine() once, this would fail, right? Maybe that should be added to a comment?

Reply:
Good point! I think it makes sense to advance the buffer when nothing has been read, wdyt?

Reply:
Sounds good!